File: kor-common/src/main/java/com/sefford/kor/interactors/OptimisticNetworkInteractor.java
package com.sefford.kor.interactors;

import com.sefford.common.interfaces.Loggable;
import com.sefford.common.interfaces.Postable;
import com.sefford.kor.errors.Error;
import com.sefford.kor.interactors.interfaces.Notifiable;
import com.sefford.kor.interactors.interfaces.OptimisticDelegate;
import com.sefford.kor.responses.Response;

/**
 * Standard Network Interactor that performs the Request Process in this order:
 * * <ul>
 * <li>Performing Changes to Cache</li>
 * <li>Network Retrieval phase.</li>
 * <li>Post processing.</li>
 * <li>Notifying to the UI for success</li>
 * <p/>
 * </ul>
 * <p/>
 * In any moment the Strategy can notify of an error through {@link Notifiable#notify(Object) NotifyError} interface.
 *
 * @author Saul Diaz <sefford@gmail.com>
 */
public class OptimisticNetworkInteractor<R extends Response, E extends Error> extends Interactor<R, E> {

    /**
     * Creates a new instance of Base Network Interactor
     *
     * @param bus      Notification Facility
     * @param log      Logging facilities
     * @param delegate Request to execute
     */
    public OptimisticNetworkInteractor(Postable bus, Loggable log, OptimisticDelegate<R, E> delegate) {
        super(bus, log, delegate);
    }

    @Override
    public Object execute() {
        try {
            ((OptimisticDelegate<R, E>) delegate).performCacheChanges();
            final R content = ((OptimisticDelegate<R, E>) delegate).execute();
            final R processedContent = ((OptimisticDelegate<R, E>) delegate).postProcess(content);
            long start = System.currentTimeMillis();
            log.d(TAG, delegate.getInteractorName() + "(Saving):" + (System.currentTimeMillis() - start) + "ms");
            return processedContent;
        } catch (Exception x) {
            final E error = ((OptimisticDelegate<R, E>) delegate).composeErrorResponse(x);
            ((OptimisticDelegate<R, E>) delegate).revertCacheStatus();
            if (loggable && error.isLoggable()) {
                delegate.logErrorResponse(log, x);
            }
            return error;
        }
    }
}

File: modules/kor-usecases/src/main/kotlin/com/sefford/kor/interactors/OptimisticNetworkInteractor.kt
package com.sefford.kor.interactors

import com.sefford.common.interfaces.Loggable
import com.sefford.common.interfaces.Postable
import com.sefford.kor.interactors.delegates.Notifiable
import com.sefford.kor.interactors.delegates.OptimisticDelegate
import com.sefford.kor.usecases.components.Response
import com.sefford.kor.usecases.components.Error
/**
 * Standard Network Interactor that performs the Request Process in this order:
 * *
 *  * Performing Changes to Cache
 *  * Network Retrieval phase.
 *  * Post processing.
 *  * Notifying to the UI for success
 *
 *
 *
 *
 *
 * In any moment the Strategy can notify of an error through [NotifyError][Notifiable.notify] interface.
 *
 * @author Saul Diaz <sefford@gmail.com>
 */
class OptimisticNetworkInteractor<R : Response, E : Error>
/**
 * Creates a new instance of Base Network Interactor
 *
 * @param bus      Notification Facility
 * @param log      Logging facilities
 * @param delegate Request to execute
 */
(bus: Postable, log: Loggable, delegate: OptimisticDelegate<R, E>) : Interactor<R, E>(bus, log, delegate) {

    override fun execute(): Any {
        val start = System.currentTimeMillis()
        try {
            delegate.startPerformanceLog(log)
            (delegate as OptimisticDelegate<R, E>).performCacheChanges()
            val content = delegate.execute()
            val processedContent = delegate.postProcess(content)
            delegate.endPerformanceLog(log, System.currentTimeMillis() - start)
            return processedContent
        } catch (x: Exception) {
            val error = (delegate as OptimisticDelegate<R, E>).composeErrorResponse(x)
            delegate.revertCacheStatus()
            if (loggable && error.loggable) {
                delegate.logErrorResponse(log, x)
            }
            delegate.endPerformanceLog(log, System.currentTimeMillis() - start)
            return error
        }

    }
}
